test/coverage-test/test/configuration-options.test-d.ts

```diff
@@ -0,0 +1,133 @@
+import { assertType, test } from 'vitest'
+import type { ResolvedCoverageOptions, Vitest } from 'vitest'
+import type { defineConfig } from 'vitest/config'
+
+type NarrowToTestConfig<T> = T extends { test?: any } ? NonNullable<T['test']> : never
+type Configuration = NarrowToTestConfig<(Parameters<typeof defineConfig>[0])>
+type Coverage = NonNullable<Configuration['coverage']>
+
+test('providers, built-in', () => {
+  assertType<Coverage>({ provider: 'c8' })
+  assertType<Coverage>({ provider: 'istanbul' })
+
+  // @ts-expect-error -- String options must be known built-in's
+  assertType<Coverage>({ provider: 'unknown-reporter' })
+})
+
+test('providers, custom', () => {
+  assertType<Coverage>({
+    provider: {
+      getProvider() {
+        return {
+          name: 'custom-provider',
+          initialize(_: Vitest) {},
+          resolveOptions() {
+            return {} as ResolvedCoverageOptions
+          },
```

Comment:
Testing this requires refactoring the internal coverage typings quite much. I have it planned in a separate PR.